parlai/mturk/tasks/acute_eval/run.py

```diff
@@ -0,0 +1,413 @@
+        if subtask_data['task_specs'].get('is_onboarding', False):
+            workers_to_onboarding_tasks_todo[worker_id].append(
+                subtask_data['task_specs']['internal_pair_id']
+            )
+        else:
+            task_queue.put(subtask_data)
+            try:
+                workers_tasks_completed[worker_id].remove(
+                    subtask_data['task_specs']['internal_pair_id']
+                )
+                workers_to_conversations_seen[worker_id].remove(
+                    subtask_data['pairing_dict']['dialogue_dicts'][0]['id']
+                )
+                workers_to_conversations_seen[worker_id].remove(
+                    subtask_data['pairing_dict']['dialogue_dicts'][1]['id']
+                )
+            except ValueError():
+                print("WARNING: couldn't remove task from worker's history")
+
+
+def get_onboarding_tasks(worker_id, tasks_per_hit):
+    """
+    Get the next onboarding task for this worker id.
+
+    If the worker has never done a task, shuffle the onboarding tasks for them. If
+    they've done all of the onboarding tasks or if there are no onboarding tasks, return
+    None
+    """
+    if len(onboarding_tasks) == 0:
+        return []
+    onboarding_tasks_todo = workers_to_onboarding_tasks_todo.get(worker_id)
+    if onboarding_tasks_todo is None:
+        onboarding_tasks_todo = list(onboarding_tasks.keys())
+        np.random.shuffle(onboarding_tasks_todo)
+        workers_to_onboarding_tasks_todo[worker_id] = onboarding_tasks_todo
+    if len(onboarding_tasks_todo) == 0:
+        return []
+    num_tasks_to_return = min(len(onboarding_tasks_todo), tasks_per_hit)
+    onboarding_tasks_chosen = onboarding_tasks_todo[:num_tasks_to_return]
+    workers_to_onboarding_tasks_todo[worker_id] = onboarding_tasks_todo[
+        num_tasks_to_return:
+    ]
+    return [onboarding_tasks[id] for id in onboarding_tasks_chosen]
+
+
+def check_and_update_worker_approval(mturk_manager, worker_id, threshold, save_data):
+    """
+    Soft block workers who fail onboarding tasks, keep track of their status.
+    """
+    all_task_data = save_data['worker_data'][worker_id]['task_data']
+    response_data = save_data['worker_data'][worker_id]['response']['task_data']
+    num_onboarding_tasks = 0
+    num_correct = 0
+    for i in range(len(all_task_data)):
+        is_onboarding = all_task_data[i]['pairing_dict'].get('is_onboarding', False)
+        if not is_onboarding:
+            continue
+        worker_response = response_data[i]['speakerChoice']
+        expected_response = all_task_data[i]['pairing_dict']['correct_answer']
+        num_onboarding_tasks += 1
+        if worker_response == expected_response:
+            num_correct += 1
+    if num_onboarding_tasks == 0:
+        if worker_id in onboarding_failed_workers:
+            return_task_data(worker_id, all_task_data)
+        return
+    if (num_correct / num_onboarding_tasks) >= threshold:
```

Comment:
can you add some  more comments in the code explaining the flow through this funciton?